rlp: Use BytesMut for RlpStream's backing buffer

vorot93 opened the pull request:
While profiling rust-devp2p, I've noticed that a noticeable portion of allocations comes from `rlp`. I suspect that this comes from rlp allocating byte buffer for each `RlpStream` - even though it is really a mere writer.

This PR switches `RlpStream` to accept user-provided `bytes::BytesMut` as internal buffer.
* In my case (building RLPx packets) this allows for completely reusing `BytesMut` provided by `tokio`'s Framed, for potentially zero allocations for each packet.
* `bytes` is a pretty established crate at this point. It is no_std ready and is used all across Tokio ecosystem - from Tokio itself all the way to hyper.

ordian commented on rlp/src/stream.rs:
```diff
@@ -26,43 +27,39 @@ impl ListInfo {
 	/// Initializes instance of empty `Stream`.
-	pub fn new() -> Self {
-		RlpStream { unfinished_lists: Vec::with_capacity(16), buffer: Vec::with_capacity(1024), finished_list: false }
+	pub fn new(buffer: &'a mut BytesMut) -> Self {
```
I'm not a fan of tying this lib to only `BytesMut`, is it possible to make it generic so that it can accept both a vec and bytesmut?

vorot93 replied:
There is `BufMut` trait, but it is rather abridged in that it does not provide methods for controlling allocation and determining current buffer length. Besides, trying to make `RlpStream` generic would break `Encodable`'s object safety.

ordian replied:
Ok, I see.

ordian reviewed (commented):
I'd try to make this change non-breaking by keeping existing `new`, `new_list` methods and adding `new_with_buffer`, `new_list_with_buffer` methods and making `RlpStream` generic on buffer with the default on `Vec` if that route is possible.

ordian commented on rlp/Cargo.toml:
```diff
@@ -8,6 +8,7 @@ authors = ["Parity Technologies <admin@parity.io>"]
 edition = "2018"
 
 [dependencies]
+bytes = "0.6"
```
we should put `bytes/std` behind `std` feature

vorot93 replied:
fixed

vorot93 commented:
@ordian I understand that this PR hurts ergonomics for cases where there is no preallocated buffer. However, I'm not sure how to genericize over owned and mutably borrowed `BytesMut` without breaking `Encodable` 😕

ordian commented on rlp/src/stream.rs:
````diff
@@ -231,36 +230,18 @@ impl RlpStream {
-	/// let out = stream.out();
-	/// assert_eq!(out, vec![0x83, b'd', b'o', b'g']);
-	/// ```
-	pub fn clear(&mut self) {
````
why is this method removed? we can clear `self.buffer[self.start_pos..]`

ordian reviewed (approved):
I'm ok with this change, but let's see what others think.

vorot93 commented:
After tinkering a bit more with BytesMut API it looks like we can still do it with owned buffer **and** avoid extra allocations.

I have redone the PR and added the test that shows how to use `BytesMut::split`.

ordian reviewed (approved):
Thanks! That's still a breaking change, but looks better now.

dvdplm reviewed (approved):
We switched to `Vec` from using `ElasticArray` a while back after benchmarking showed no regressions, but it's possible that the benchmarks were flawed and did not reflect real world usage. Either way, my point is that the the impact on the ecosystem was not very big – there was one messy PR to OpenEthereum and that was it I think.
I think it's fair to prioritize performance over avoiding breakage for this crate. LGTM.

(FWIW I think that the initialization buffer sizes used in this crate were always "guesstimates" and that they could be tweaked)

dvdplm commented:
@niklasad1 any thoughts on this before we merge?

niklasad1 commented on rlp/src/stream.rs:
```diff
@@ -71,11 +82,6 @@ impl RlpStream {
 	}
 
-	/// Drain the object and return the underlying ElasticArray. Panics if it is not finished.
-	pub fn drain(self) -> Vec<u8> {
```
This is a breaking change as I suppose?

ordian replied:
That's not the only breaking change (but it doesn't really matter, because this function is an alias to another public function). Every funtion that returned `Vec` now returns `BytesMut`.

niklasad1 reviewed (approved):
nice work, LGTM
